File: docs/js/balcony-smoke.js
/**
 * Suite balcony POV smoke-break minigame — session ledger + access gates.
 * Web overlay: BalconySmokeOverlay.js
 */

import { ensureHotel, getRoomType } from "./hotel.js";
import { recordConsumption, getIntoxicationSummary } from "./intoxication-effects.js";
import { canAccessHotelRoom } from "./world-cycle.js";

/** Consumable recorded on each hit from the suite balcony. */
export const BALCONY_JOINT_ID = "balcony_suite_joint";

export const BALCONY_SMOKE_ROOM_TYPES = ["suite", "penthouse"];

export const BALCONY_HIT_MAX = 5;

export const BALCONY_VIBE_LINES = [
  "Luxor’s beam cuts the haze. You own this altitude for one more inhale.",
  "Traffic crawls like chips across a felt. You are above the action.",
  "Neon blooms. The Strip performs for anyone with a key this high.",
  "A warm wind lifts the ember. Somewhere below, someone loses a bet.",
  "Bellagio’s fountains stutter in the distance. You don’t hurry for anyone.",
  "The glass railing holds a reflection of a whale who finally sat still.",
  "City lights smear into gold. High roller stillness. Perfect.",
];

export function defaultBalconySmokeState(overrides = {}) {
  return {
    visits: overrides.visits ?? 0,
    lifetimeHits: overrides.lifetimeHits ?? 0,
    lastVisitAt: overrides.lastVisitAt ?? null,
    eggs: [...(overrides.eggs ?? [])],
  };
}

export function ensureBalconySmoke(session) {
  if (!session.balconySmoke) {
    session.balconySmoke = defaultBalconySmokeState();
  }
  const defaults = defaultBalconySmokeState();
  for (const key of Object.keys(defaults)) {
    if (session.balconySmoke[key] === undefined) {
      session.balconySmoke[key] = defaults[key];
    }
  }
  if (!Array.isArray(session.balconySmoke.eggs)) {
    session.balconySmoke.eggs = [];
  }
  return session.balconySmoke;
}

export function attachBalconySmokeToSession(session, data) {
  if (data?.balconySmoke) {
    session.balconySmoke = defaultBalconySmokeState(data.balconySmoke);
  }
}

/**
 * @param {import("./core.js").PlayerSession} session
 * @returns {{ ok: boolean, message?: string, roomLabel?: string }}
 */
export function canEnterBalconySmoke(session) {
  const hotel = ensureHotel(session);
  if (!canAccessHotelRoom(session) || !hotel.reachedRoom) {
    return { ok: false, message: "Reach your suite door first — then the balcony is yours." };
  }
  if (!BALCONY_SMOKE_ROOM_TYPES.includes(hotel.roomType)) {
    return {
      ok: false,
      message: "Strip POV smoke breaks are a suite and penthouse perk. Ask Carmen about an upgrade.",
    };
  }
  const room = getRoomType(hotel);
  return { ok: true, roomLabel: room.label };
}

export function startBalconyVisit(session) {
  const gate = canEnterBalconySmoke(session);
  if (!gate.ok) return { ok: false, message: gate.message };
  const ledger = ensureBalconySmoke(session);
  ledger.visits += 1;
  ledger.lastVisitAt = Date.now();
  return {
    ok: true,
    roomLabel: gate.roomLabel,
    visits: ledger.visits,
  };
}

/**
 * Take a hit — records intoxication and advances the sitting.
 * @returns {{ ok: boolean, message: string, hits: number, done: boolean, vibe: string, intoxLevel: number }}
 */
export function takeBalconyHit(session, sitting) {
  if (!sitting || sitting.closed) {
    return { ok: false, message: "Step back onto the balcony first.", hits: 0, done: true, vibe: "", intoxLevel: 0 };
  }
  if (sitting.hits >= BALCONY_HIT_MAX) {
    return {
      ok: false,
      message: "The joint is ash. Savor the view, then step inside.",
      hits: sitting.hits,
      done: true,
      vibe: pickVibe(sitting),
      intoxLevel: getIntoxicationSummary(session).level,
    };
  }

  const r = recordConsumption(session, BALCONY_JOINT_ID, { source: "balcony_smoke" });
  sitting.hits += 1;
  const ledger = ensureBalconySmoke(session);
  ledger.lifetimeHits += 1;

  if (sitting.hits >= 3 && !ledger.eggs.includes("high_roller_haze")) {
    ledger.eggs.push("high_roller_haze");
  }

  const vibe = pickVibe(sitting);
  const summary = getIntoxicationSummary(session);
  const done = sitting.hits >= BALCONY_HIT_MAX;
  const message = r.ok
    ? (done
      ? `Final hit. ${vibe}`
      : `Hit ${sitting.hits}/${BALCONY_HIT_MAX}. ${vibe}`)
    : "The ember catches — atmosphere only, but the Strip still performs.";

  return {
    ok: true,
    message,
    hits: sitting.hits,
    done,
    vibe,
    intoxLevel: summary.level,
  };
}

export function createBalconySitting(session) {
  const hotel = ensureHotel(session);
  return {
    roomType: hotel.roomType,
    roomNumber: hotel.roomNumber,
    floor: hotel.floor,
    hits: 0,
    closed: false,
    startedAt: Date.now(),
  };
}

export function closeBalconySitting(session, sitting) {
  if (!sitting || sitting.closed) {
    return { ok: true, message: "Back inside. The carpet smells like victory and HVAC." };
  }
  sitting.closed = true;
  const ledger = ensureBalconySmoke(session);
  const hits = sitting.hits;
  if (hits === 0) {
    return { ok: true, message: "You watched the Strip without lighting up. Discipline — or just pacing yourself." };
  }
  if (hits >= BALCONY_HIT_MAX) {
    return {
      ok: true,
      message: `Session complete — ${hits} hits from floor ${sitting.floor}. Lifetime balcony hits: ${ledger.lifetimeHits}.`,
    };
  }
  return {
    ok: true,
    message: `You step inside after ${hits} hit(s). The Strip keeps glittering without you.`,
  };
}

function pickVibe(sitting) {
  const idx = (sitting.hits + (sitting.roomNumber % 7)) % BALCONY_VIBE_LINES.length;
  return BALCONY_VIBE_LINES[idx];
}

File: mandalay_bay/balcony_smoke.py
"""Suite balcony POV smoke-break ledger (CLI parity for web overlay)."""

from __future__ import annotations

from dataclasses import dataclass, field

from mandalay_bay.hotel import ensure_hotel
from mandalay_bay.intoxication import record_consumption
from mandalay_bay.session import PlayerSession
from mandalay_bay.world_cycle import can_access_hotel_room

BALCONY_JOINT_ID = "balcony_suite_joint"
BALCONY_SMOKE_ROOM_TYPES = ("suite", "penthouse")
BALCONY_HIT_MAX = 5

BALCONY_VIBE_LINES = [
    "Luxor's beam cuts the haze. You own this altitude for one more inhale.",
    "Traffic crawls like chips across a felt. You are above the action.",
    "Neon blooms. The Strip performs for anyone with a key this high.",
    "A warm wind lifts the ember. Somewhere below, someone loses a bet.",
    "Bellagio's fountains stutter in the distance. You don't hurry for anyone.",
]


@dataclass
class BalconySmokeState:
    visits: int = 0
    lifetime_hits: int = 0
    last_visit_at: int | None = None
    eggs: list[str] = field(default_factory=list)


@dataclass
class BalconySitting:
    room_type: str
    room_number: int
    floor: int
    hits: int = 0
    closed: bool = False


@dataclass
class ActionResult:
    ok: bool
    message: str
    hits: int = 0
    done: bool = False


def ensure_balcony_smoke(session: PlayerSession) -> BalconySmokeState:
    if not hasattr(session, "balcony_smoke") or session.balcony_smoke is None:
        session.balcony_smoke = BalconySmokeState()
    return session.balcony_smoke


def can_enter_balcony_smoke(session: PlayerSession) -> ActionResult:
    hotel = ensure_hotel(session)
    if not can_access_hotel_room(session) or not hotel.reached_room:
        return ActionResult(False, "Reach your suite door first — then the balcony is yours.")
    if hotel.room_type not in BALCONY_SMOKE_ROOM_TYPES:
        return ActionResult(
            False,
            "Strip POV smoke breaks are a suite and penthouse perk. Ask Carmen about an upgrade.",
        )
    return ActionResult(True, "Balcony open.")


def start_balcony_visit(session: PlayerSession) -> tuple[ActionResult, BalconySitting | None]:
    gate = can_enter_balcony_smoke(session)
    if not gate.ok:
        return gate, None
    hotel = ensure_hotel(session)
    ledger = ensure_balcony_smoke(session)
    ledger.visits += 1
    sitting = BalconySitting(
        room_type=hotel.room_type,
        room_number=hotel.room_number,
        floor=hotel.floor,
    )
    return ActionResult(True, f"Floor {hotel.floor} balcony — visit #{ledger.visits}."), sitting


def take_balcony_hit(session: PlayerSession, sitting: BalconySitting) -> ActionResult:
    if sitting.closed:
        return ActionResult(False, "Step back onto the balcony first.", done=True)
    if sitting.hits >= BALCONY_HIT_MAX:
        return ActionResult(False, "The joint is ash. Savor the view, then step inside.", hits=sitting.hits, done=True)
    record_consumption(session, BALCONY_JOINT_ID, source="balcony_smoke")
    sitting.hits += 1
    ledger = ensure_balcony_smoke(session)
    ledger.lifetime_hits += 1
    if sitting.hits >= 3 and "high_roller_haze" not in ledger.eggs:
        ledger.eggs.append("high_roller_haze")
    vibe = BALCONY_VIBE_LINES[(sitting.hits + sitting.room_number) % len(BALCONY_VIBE_LINES)]
    done = sitting.hits >= BALCONY_HIT_MAX
    msg = f"Final hit. {vibe}" if done else f"Hit {sitting.hits}/{BALCONY_HIT_MAX}. {vibe}"
    return ActionResult(True, msg, hits=sitting.hits, done=done)


def close_balcony_sitting(session: PlayerSession, sitting: BalconySitting) -> ActionResult:
    if sitting.closed:
        return ActionResult(True, "Already back inside.")
    sitting.closed = True
    ledger = ensure_balcony_smoke(session)
    if sitting.hits == 0:
        return ActionResult(True, "You watched the Strip without lighting up.")
    return ActionResult(
        True,
        f"You step inside after {sitting.hits} hit(s). Lifetime balcony hits: {ledger.lifetime_hits}.",
        hits=sitting.hits,
    )
